controllers/secret_controller.go

```diff
@@ -217,42 +151,82 @@ func (r *ReconcileSecret) Reconcile(request reconcile.Request) (reconcile.Result
+func (r *SecretReconciler) RemoveInvalidAnnotationsFromLinuxNodes(config *rest.Config) {
+	// create a new clientset as this function will be called before the manager's client is started
+	kc, err := kubernetes.NewForConfig(config)
 	if err != nil {
-		return errors.Wrapf(err, "error getting node list")
+		r.log.Error(err, "error creating kubernetes clientset")
+	}
+
+	nodes, err := kc.CoreV1().Nodes().List(context.TODO(), meta.ListOptions{LabelSelector: core.LabelOSStable + "=linux"})
+	if err != nil {
+		r.log.Error(err, "error getting node list")
 	}
 	// The public key hash was accidentally added to Linux nodes in WMCO 2.0 and must be removed.
 	// The `/` in the annotation key needs to be escaped in order to not be considered a "directory" in the path.
 	escapedPubKeyAnnotation := strings.Replace(nodeconfig.PubKeyHashAnnotation, "/", "~1", -1)
 	patchData := fmt.Sprintf(`[{"op":"remove","path":"/metadata/annotations/%s"}]`, escapedPubKeyAnnotation)
 	for _, node := range nodes.Items {
 		if _, present := node.Annotations[nodeconfig.PubKeyHashAnnotation]; present == true {
-			err = r.client.Patch(context.TODO(), &node, client.RawPatch(kubeTypes.JSONPatchType, []byte(patchData)))
+			_, err = kc.CoreV1().Nodes().Patch(context.TODO(), node.GetName(), kubeTypes.JSONPatchType,
+				[]byte(patchData), meta.PatchOptions{})
 			if err != nil {
-				return errors.Wrapf(err, "error removing public key annotation from node %s", node.GetName())
+				r.log.Error(err, "error removing public key annotation from node", "node", node.GetName())
 			}
 		}
 	}
-	return nil
-}
-
-// userDataMapper is a simple implementation of the Mapper interface allowing for the mapping from the userData secret
-// to the private key secret
-type userDataMapper struct {
-	// watchNamespace is the namespace the operator is watching as defined by the CSV
-	watchNamespace string
 }
 
-// newUserDataMapper returns a pointer to a new userDataMapper
-func newUserDataMapper(namespace string) *userDataMapper {
-	return &userDataMapper{watchNamespace: namespace}
-}
-
-func (m *userDataMapper) Map(_ handler.MapObject) []reconcile.Request {
+// MapToPrivateKeySecret is a mapping function that will always return a request for the cloud private key secret
+func MapToPrivateKeySecret(_ client.Object) []reconcile.Request {
 	return []reconcile.Request{
-		{NamespacedName: kubeTypes.NamespacedName{Namespace: m.watchNamespace, Name: secrets.PrivateKeySecret}},
+		{NamespacedName: kubeTypes.NamespacedName{Namespace: watchNamespace, Name: secrets.PrivateKeySecret}},
+	}
+}
+func (r *SecretReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// Check that the private key exists, if it doesn't, log a warning
+	_, err := secrets.GetPrivateKey(kubeTypes.NamespacedName{Namespace: r.watchNamespace, Name: secrets.PrivateKeySecret}, mgr.GetClient())
+	if err != nil {
+		r.log.Error(err, "Unable to retrieve private key, please ensure it is created")
```

Comment:
should we generate warning events for this?

Reply:
I'd rather make as little changes as possible, we can do that in a different PR if wanted.